test/operators/test_spin_op.py

```diff
@@ -0,0 +1,137 @@
+
+    def test_init_pm_label(self):
+        """Test __init__ with plus and minus label"""
+        with self.subTest("plus"):
+            plus = SpinOp([("+_0", 2)])
+            desired = SpinOp([("X_0", 2), ("Y_0", 2j)])
+            self.assertListEqual(plus.to_list(), desired.to_list())
+
+        with self.subTest("minus"):
+            minus = SpinOp([("-_0", 2)])
+            desired = SpinOp([("X_0", 2), ("Y_0", -2j)])
+            self.assertListEqual(minus.to_list(), desired.to_list())
+
+        with self.subTest("plus tensor minus"):
+            actual = SpinOp([("+_1 -_0", 3)])
+            desired = SpinOp([("X_1 X_0", 3), ("X_1 Y_0", -3j), ("Y_1 X_0", 3j), ("Y_1 Y_0", 3)])
+            self.assertSetEqual(frozenset(actual.to_list()), frozenset(desired.to_list()))
+
+    @data(*spin_labels(1), *spin_labels(2))
+    def test_init_dense_label(self, label):
+        """Test __init__ for label_mode=dense"""
+        if len(label) == 1:
+            actual = SpinOp([(f"{label}", 1 + 1j)], label_mode="dense")
+            desired = SpinOp([(f"{label}_0", 1 + 1j)])
+        elif len(label) == 2:
+            actual = SpinOp([(f"{label}", 1)], label_mode="dense")
+            desired = SpinOp([(f"{label[0]}_1 {label[1]}_0", 1)])
+        self.assertListEqual(actual.to_list(), desired.to_list())
+
+    def test_neg(self):
+        """Test __neg__"""
+        actual = -self.heisenberg
+        desired = -self.heisenberg_mat
+        np.testing.assert_array_almost_equal(actual.to_matrix(), desired)
+
+    def test_mul(self):
+        """Test __mul__, and __rmul__"""
+        actual = self.heisenberg * 2
+        np.testing.assert_array_almost_equal(actual.to_matrix(), self.heisenberg_mat * 2)
+
+    def test_div(self):
+        """Test __truediv__"""
+        actual = self.heisenberg / 3
+        np.testing.assert_array_almost_equal(actual.to_matrix(), self.heisenberg_mat / 3)
+
+    def test_add(self):
+        """Test __add__"""
+        actual = self.heisenberg + self.heisenberg
+        np.testing.assert_array_almost_equal(actual.to_matrix(), self.heisenberg_mat * 2)
+
+    def test_sub(self):
+        """Test __sub__"""
+        actual = self.heisenberg - self.heisenberg
+        np.testing.assert_array_almost_equal(actual.to_matrix(), np.zeros((9, 9)))
+
+    def test_adjoint(self):
+        """Test adjoint method and dagger property"""
+        actual = ~self.heisenberg
+        np.testing.assert_array_almost_equal(actual.to_matrix(), self.heisenberg_mat.conj().T)
+
+    def test_reduce(self):
+        """Test reduce"""
+        actual = (self.heisenberg - self.heisenberg).reduce()
+        self.assertListEqual(actual.to_list(), [("I_1 I_0", 0)])
+
+    def test_to_matrix(self):
+        """Test to_matrix()"""
+        actual = SpinOp([("X_2 Y_1 Z_0", 1)]).to_matrix()
+        desired = Pauli("XYZ").to_matrix() / 8
+        np.testing.assert_array_almost_equal(actual, desired)
+
```

Comment:
Please add tests to separately test the parsing logic. That is very complex and we should make sure it does not break.

Reply:
end2end `parsing_logic` was tested [here](https://github.com/Qiskit/qiskit-nature/pull/35/files#diff-e1fad0f243ef03387da83f3f9c9b006212b413a01addd08569d3d78048469dd1R60-R98).
The parse function cannot be used independently. What tests do you think is necessary?

Reply:
I was thinking that it may make sense to explicitly test the parsing methods like `_flatten_ladder_ops`, `_parse_dense_labels` and `_parse_sparse_labels`. Currently, all of these are always tested within the complete flow but if that test fails for some reason we have multiple places to track down. Thus, if we have single tests for each of these methods we can more easily see which part fails :+1:

Reply:
I see. I have added a test for `_flatten_ladder_ops`, but I cannot make tests for `_parse_dense_labels` and `_parse_sparse_labels ` because it's not pure functions.

Reply:
I am sorry but I don't fully understand what you mean by "not pure functions"

Reply:
They don't have return value but side effects.
https://en.wikipedia.org/wiki/Pure_function

Reply:
Aha I see. But surely you can just test that the side effects are what you expect? I don't see a problem with that

Reply:
It cannot be called by itself. They only have side effects (assign to instance attributes). 
Since `self` is required to call them, they are necessary to call `__init__`, and they are essentially same with the `__init__` tests.

Reply:
How about the following:
1. construct an empty operator: `spin_op = SpinOp(<whatever needs to go here in that case>)`
2. `spin_op._parse_dense_label(['XX'])`
3. assert the new contents of `spin_op`

And then something analogous for the sparse label parser

Reply:
In general, tests are not necessary for private methods.

In this case, these methods cannot be called by itself. The purpose of these methods is to improve readability. Even if tests are not enough we should add tests for public methods. If tests are needed here, I'll move them to `__init__`.

Reply:
I think it doesn't hurt to test private methods but if you feel so strong about this, we can leave it as is. I agree that in principle these methods will never be called by a user as is, but since we do have them split up for readability in my mind it makes sense to attempt to test them separately as well, as it allows a more fine-grained testing of the code.

But you can leave it as is.

Reply:
The goal of unit test is to test the module via its public API. When doing so it should cover testing the private methods via use of appropriate test values to drive it across the set of paths to cover all the code so they are used/tested in the context of being called by the module code via the public API since that is how the code will be used in practice in the system. Normally one does not directly test private methods, but of course the code needs to be covered by unit tests. You will find discussions on testing private methods out there so I am not going to say more